### DISCUSSION TITLE
Doubt negative sampling for link prediction

### DISCUSSION THREAD

**Discussion by yorard123** [pyg-team/pytorch_geometric#3384] [Q&A] [answered]
Hi,

I'm trying to do a negative sampling for a link prediction problem. My problem has two kinds of nodes, type1 and type2. The edges **only** can be between type1 and type2. I have been checking the documentation and I found that the current methods for batched negative sampling does not support this use case. Is there anyway in pytorch geometric to solve this issue? I was thinking to do it manually creating manually an all to all edge_index vector, and pick randomly certain number of edges, however, as I did it was very slow. I would appreciate any hint on this issue.

EDIT: Another requirement that I have is that I need that each node of type1 has the same number of edges.

**Comment by rusty1s** [comment, marked as answer]
Thanks for pointing this out. I made some changes in master, and you can now perform bipartite batch-wise negative sampling as well, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/utils/negative_sampling.py#L94-L124). This is achieved via a tuple of batch vectors as input, i.e.:
```python
batched_negative_sampling(edge_index, (src_batch, dst_batch))
```
I will craft a new release soon.

**Reply by yorard123** [reply]
Hi,

Thanks! I have one doubt related to the src_batch and dst_batch. At this moment what I have is a data structure that contains all the nodes in there, and to know whicch are of type 1 or type 2 I have a mask that it helps me to know that. For instance:
`data.x = [x, x, x, y, y, y, x, x, x, y, y]
data.batch = [0,0,0,0,0,0,1,1,1,1,1]
data.mask_1 = [1,1,1,0,0,0,1,1,1,0,0]`

Assume that x is type 1 and y type 2. My question is, what is expected to be the src_batch and the dst_batch?

Thanks!

**Reply by rusty1s** [reply]
`src_batch` and `dst_batch` refer to mini-batching with `HeteroData` objects, which is currently the recommended way to operate on heterogeneous graphs. That is, different node types are stored in distinct matrices, and are connected via bipartite graphs. For your use-case, it's probably best to simply over-sample the amount of negative edges, and then only keep the ones that connect different types:
```python
neg_edge_index = batched_negative_sampling(edge_index, batch, ...)
mask = data.mask_1[edge_index[0]] & ~(data.mask_1[edge_index[1]])
neg_edge_index = neg_edge_index[:, mask]
```

**Reply by yorard123** [reply]
Thanks!
